File: algorithm/leetcode68.text-justification.cpp
class Solution {
public:
    vector<string> fullJustify(vector<string>& words, int maxWidth) {
        vector<string> ans;
        string line;
        int i, width = 0, wc;
        for(i = 0; i < words.size(); ++i)
        {
        	if(line.empty())
        	{
        		line = words[i];
        		width = words[i].size();
        		wc = 1;
        	}
        	else
        	{
        		if(width+1+words[i].size() <= maxWidth)
        		{
        			line += " "+words[i];
        			width += 1+words[i].size();
        			// wordwith += words[i].size();
        			wc++;
        		}
        		else//超了，放不下i
        		{
        			process(wc,line,maxWidth,width);
        			ans.push_back(line);
    				line = "";
    				width = wc = 0;
    				i--;
        		}
        	}
        }
        line += string(maxWidth-width,' ');
        ans.push_back(line);
        return ans;
    }

    void process(int wc, string& line, int maxWidth, int width)
    {
    	if(wc == 1)
    	{
    		line += string(maxWidth-width,' ');
    		return;
    	}
    	int space = maxWidth - width;
    	int n = space/(wc-1);
    	int pos = wc-1;
    	for(int i = line.size()-1; i >= 0; --i)
    	{
    		if(line[i] == ' ')
    		{
    			line.insert(i,n,' ');
    			space -= n;
    			pos--;
    			if(pos > 0 && space%pos == 0)
    				n = space/pos;
    		}
    	}
    }
};

File: algorithm/Python3_version/leetcode68.text-justification.py
class Solution:
    def fullJustify(self, words: List[str], maxWidth: int) -> List[str]:
        ans = []
        line = ""
        width = 0
        wc = 0
        def process(wc,line,width):
            if wc==1:
                line += ' '*(maxWidth-width)
                return line
            space = maxWidth-width
            n = space//(wc-1)
            pos = wc-1
            line = list(line)
            size = len(line)
            for i in range(size-1,-1,-1):
                if line[i]==' ':
                    line.insert(i, ' '*n)
                    space -= n
                    pos -= 1
                    if pos > 0 and space%pos==0:
                        n = space//pos
            line = "".join(line)
            return line
        i = 0
        while i < len(words):
            if len(line)==0:
                line = words[i]
                width = len(words[i])
                wc = 1
            else:
                if width+1+len(words[i]) <= maxWidth:
                    line += " "+words[i]
                    width += 1+len(words[i])
                    wc += 1
                else:
                    temp = process(wc,line,width)
                    ans.append(temp)
                    line = ""
                    width, wc = 0, 0
                    i -= 1
            i += 1
        line += ' '*(maxWidth-width)
        ans.append(line)
        return ans